src/georaster-layer-for-leaflet.ts

```diff
@@ -0,0 +1,592 @@
+import getProjString from "utm-utils/src/getProjString";
+import type { Coords, DoneCallback, LatLngTuple } from "leaflet";
+import type {
+  GeorasterLayerOptions,
+  Georaster,
+  GeorasterKeys,
+  GetRasterOptions,
+  DrawTileOptions,
+  PixelValueToColorFn,
+  Tile
+} from "./types";
+
+const EPSG4326 = 4326;
+const PROJ4_SUPPORTED_PROJECTIONS = new Set([3857, 4269]);
+const MAX_NORTHING = 1000;
+const MAX_EASTING = 1000;
+const ORIGIN: LatLngTuple = [0, 0];
+
+const GeoRasterLayer = L.GridLayer.extend({
+  initialize: function (options: GeorasterLayerOptions) {
+    try {
+      if (options.georasters) {
+        this.georasters = options.georasters;
+      } else if (options.georaster) {
+        this.georasters = [options.georaster];
+      } else {
+        throw new Error("You initialized a GeoRasterLayer without a georaster or georasters value.");
+      }
+
+      /*
+          Unpacking values for use later.
+          We do this in order to increase speed.
+      */
+      const keys = [
+        "height",
+        "width",
+        "noDataValue",
+        "palette",
+        "pixelHeight",
+        "pixelWidth",
+        "projection",
+        "sourceType",
+        "xmin",
+        "xmax",
+        "ymin",
+        "ymax"
+      ];
+      if (this.georasters.length > 1) {
+        keys.forEach(key => {
+          if (this.same(this.georasters, key)) {
+            this[key] = this.georasters[0][key];
+          } else {
+            throw new Error("all GeoRasters must have the same " + key);
+          }
+        });
+      } else if (this.georasters.length === 1) {
+        keys.forEach(key => {
+          this[key] = this.georasters[0][key];
+        });
+      }
+
+      // used later if simple projection
+      this.ratio = this.height / this.width;
+
+      if (this.sourceType === "url") {
+        if (!options.updateWhenIdle) options.updateWhenIdle = false;
+        if (!options.updateWhenZooming) options.updateWhenZooming = true;
+        if (!options.keepBuffer) options.keepBuffer = 16;
+      }
+
+      // todo: set defaults in descructured options - SFR 2021-01-20
+      if (!("debugLevel" in options)) options.debugLevel = 1;
+      if (!options.keepBuffer) options.keepBuffer = 25;
+      if (!options.resolution) options.resolution = Math.pow(2, 5);
+      if (options.updateWhenZooming === undefined) options.updateWhenZooming = false;
+
+      this.debugLevel = options.debugLevel;
+      if (this.debugLevel >= 1) console.log("georaster:", options);
+
+      //! note: Values is missing in type. What should it be? number[][][] - SFR 2021-01-20
+      if (this.georasters.every((georaster: Georaster) => typeof georaster.values === "object")) {
+        this.rasters = this.georasters.reduce((result: number[], georaster: Georaster) => {
+          result = result.concat(georaster.values);
+          return result;
+        }, []);
+        if (this.debugLevel > 1) console.log("this.rasters:", this.rasters);
+      }
```

Comment:
@DanielJDufour, This block is confusing me a bit. You mentioned that `georaster.values` could be undefined if the `source` is `url`. The compiler is complaining on line 88 that the concat cannot have an `| undefined` type. Is there a different way we could write this so that we are sure `values` is defined?

Reply:
Would it be possible to do 

`result.concat(georaster.values || []);`

or:
```js
        this.rasters = this.georasters.reduce((result: number[], georaster: Georaster) => {
          if (georaster.values) result = result.concat(georaster.values);
          return result;
        }, []);
```

Reply:
You're correct that values should be `number[][][] `.  It's a three dimensional array of numbers.  The first dimension is band, second is row, and third is column.

Reply:
But it could also be undefined if georaster isn't loading all the values into memory.

Reply:
Neither of those worked I'm afraid. What is `this.rasters` expected to be? Is it a four dimensional array given the `georaster.values` is a `number[][][]`

Could it be rewritten as: 

```typescript
 this.rasters = this.georasters.map((georaster: Georaster) => georaster.values );
```

Effectively pulling all the values from the georaster into an array.

Reply:
I tried my solution above and it didn't work until I changed it to: 

```javascript
this.rasters = this.georasters.map((georaster: Georaster) => georaster.values )[0];
```

It's confusing me a bit why that necessary but it seems to work. I'm wary of changing the implementation but would that do what you expected it too?

Reply:
I believe this would return the values for the first georaster only.  For example, with Landsat8 imagery on AWS, each band is its own GeoTIFF file.  In this case, the above code would only pull the first file.

To get around the compiler/signature issues could we default the value to null in the code and update the signature to `null | [][][]`?  Seems like defaulting values to null instead of undefined is a way to get around undefined issues.